src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -595,14 +602,109 @@ CreateShardsOnWorkers(Oid distributedRelationId, List *shardPlacements,
  */
-void
-WorkerCreateShard(Oid relationId, int shardIndex, uint64 shardId, List *ddlCommandList,
-				  List *foreignConstraintCommandList, MultiConnection *connection)
+static List *
+RelationShardListForShardCreate(ShardInterval *shardInterval)
 {
+	List *relationShardList = NIL;
+	RelationShard *relationShard = NULL;
+	Oid relationId = shardInterval->relationId;
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	List *referencedRelationList = cacheEntry->referencedRelationsViaForeignKey;
+	List *referencingRelationList = cacheEntry->referencingRelationsViaForeignKey;
+	List *allForeignKeyRelations =
+		list_concat_unique_oid(referencedRelationList, referencingRelationList);
+	int shardIndex = -1;
+
+	/* record the placement access of the shard itself */
+	relationShard = CitusMakeNode(RelationShard);
+	relationShard->relationId = relationId;
+	relationShard->shardId = shardInterval->shardId;
+	relationShardList = list_make1(relationShard);
+
+	if (cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH &&
+		cacheEntry->colocationId != INVALID_COLOCATION_ID)
+	{
+		shardIndex = ShardIndex(shardInterval);
+	}
+
+	/* all foregin constraint relations */
+	if (allForeignKeyRelations != NIL)
+	{
+		ListCell *relationIdCell = NULL;
+
+		foreach(relationIdCell, allForeignKeyRelations)
+		{
+			Oid fkeyRelationid = lfirst_oid(relationIdCell);
+			RelationShard *fkeyRelationShard = NULL;
+			uint64 fkeyShardId = INVALID_SHARD_ID;
+
+			if (!IsDistributedTable(fkeyRelationid))
+			{
+				/* we're not interested in local tables */
+				continue;
+			}
+
+			if (PartitionMethod(fkeyRelationid) == DISTRIBUTE_BY_NONE)
+			{
+				fkeyShardId = GetFirstShardId(fkeyRelationid);
+			}
+			else if (cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH &&
+					 PartitionMethod(fkeyRelationid) == DISTRIBUTE_BY_HASH)
+			{
+				/* hash distributed tables should be colocated to have fkey */
+				Assert(TableColocationId(fkeyRelationid) == cacheEntry->colocationId);
+
+				fkeyShardId =
+					ColocatedShardIdInRelation(fkeyRelationid, shardIndex);
+			}
+			else
+			{
+				/*
+				 * We currently do not support foreign keys from/to local tables or
+				 * non-colocated tables when creating shards. Also note that shard
+				 * creation via shard moves doesn't happen in a transaction block,
+				 * so not relevant here.
+				 */
+				continue;
+			}
+
+			fkeyRelationShard = CitusMakeNode(RelationShard);
+			fkeyRelationShard->relationId = fkeyRelationid;
+			fkeyRelationShard->shardId = fkeyShardId;
+
+			relationShardList = lappend(relationShardList, fkeyRelationShard);
+		}
+	}
+
+	/* if  partitioned table, make sure to record the parent table */
+	if (PartitionTable(relationId))
+	{
+		RelationShard *parentRelationShard = CitusMakeNode(RelationShard);
+
+		parentRelationShard->relationId = PartitionParentOid(relationId);
+		parentRelationShard->shardId =
+			ColocatedShardIdInRelation(parentRelationShard->relationId, shardIndex);
```

Comment:
Do we support partitioned reference tables? If yes, then this line will probably have some kind of fault.

Reply:
> Do we support partitioned reference tables? If yes, then this line will probably have some kind of fault (because shardIndex is -1)

I think we don't, so maybe just add some asserts here.

Reply:
> > Do we support partitioned reference tables? If yes, then this line will probably have some kind of fault (because shardIndex is -1)
> 
> I think we don't, so maybe just add some asserts here.

Maybe just `Assert(shardIndex != -1)`.

Reply:
Makes sense added comment as well.